src/Bicep.RoslynAnalyzers/LinterRuleTypeGenerator.cs

```diff
@@ -0,0 +1,196 @@
+                .Where(method => method is not null);
+
+            var bicepAnalyzerRuleClasses = initContext.SyntaxProvider
+                .CreateSyntaxProvider(
+                    predicate: (syntax, cancellationToken) => syntax is ClassDeclarationSyntax,
+                    transform: (syntaxContext, cancellationToken) => TryGetLinterRuleContext(syntaxContext, cancellationToken))
+                .Where(ruleClass => ruleClass is not null);
+
+            var combined = generatorMethods.Combine(bicepAnalyzerRuleClasses.Collect());
+
+            initContext.RegisterSourceOutput(combined, GenerateSource);
+        }
+
+        
+        private static GeneratorMethodContext? TryGetGeneratorMethodContext(GeneratorSyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            if (syntaxContext.Node is not MethodDeclarationSyntax method)
+            {
+                // shouldn't really happen
+                return null;
+            }
+
+            var attributeSyntax = method.AttributeLists
+                .SelectMany(attributeListSyntax => attributeListSyntax.Attributes)
+                .FirstOrDefault(attributeSyntax => IsGeneratorAttribute(syntaxContext, attributeSyntax, cancellationToken));
+
+            if(attributeSyntax is null)
+            {
+                // did not find our generator attribute
+                return null;
+            }
+
+            var classSyntax = method.FirstAncestorOrSelf<ClassDeclarationSyntax>();
+            if (classSyntax is null)
+            {
+                return null;
+            }
+
+            if (syntaxContext.SemanticModel.GetDeclaredSymbol(classSyntax, cancellationToken: cancellationToken) is not INamedTypeSymbol classSymbol)
+            {
+                return null;
+            }
+
+            return new(method, classSymbol);
+        }
+
+        private static LinterRuleClassContext? TryGetLinterRuleContext(GeneratorSyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            if (syntaxContext.Node is not ClassDeclarationSyntax @class ||
+                syntaxContext.SemanticModel.GetDeclaredSymbol(@class) is not INamedTypeSymbol classSymbol)
+            {
+                // shouldn't happen
+                return null;
+            }
+
+            if(classSymbol.IsAbstract)
+            {
+                // reject abstract classes
+                return null;
+            }
+
+            var ruleTypeSymbol = syntaxContext.SemanticModel.Compilation.GetTypeByMetadataName("Bicep.Core.Analyzers.Interfaces.IBicepAnalyzerRule");
+            if(ruleTypeSymbol is null)
+            {
+                // bicep rule interface doesn't exist
+                return null;
+            }
+
+            if(!syntaxContext.SemanticModel.Compilation.ClassifyConversion(classSymbol, ruleTypeSymbol).IsImplicit)
```

Comment:
Should we not allow explicit implemenetation?

Reply:
This refers to implicit type conversion (explicit conversion would be type casting in c#). I tried an implicit interface implementation of a linter rule and it's picked up by the generator:
<img width="1497" alt="image" src="https://user-images.githubusercontent.com/22460039/192422324-4d3b2536-47e3-4198-b4ad-f054c36d1925.png">

In fact, the rule is picked up by the generator before all the methods are implemented.